### PR TITLE
Add tpcNCls and eta information for QA check

### PR DESCRIPTION
**Pull request by chengtt0406**

### CODE REVIEW

**Comment by mfaggin** [comment]
@njacazio I'd leave this to you

**Comment by njacazio** [comment]
@mfaggin  I would also involve @jezwilkinson

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 1]
```diff
@@ -386,6 +408,11 @@ struct tpcPidQa {
         // Fill histograms
         histos.fill(HIST(hexpected[id]), t.tpcInnerParam(), t.tpcSignal() - diff);
         histos.fill(HIST(hdelta[id]), t.tpcInnerParam(), diff);
+
+        if (flagFourDemensionQA) {
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:412]:
```cxx
if (fillTHnSparses) {

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 2]
```diff
@@ -404,6 +431,9 @@ struct tpcPidQa {
         const auto& nsigmatof = o2::aod::pidutils::tofNSigma<id>(t);
         if (std::abs(nsigmatof) < 3.f) {
           histos.fill(HIST(hnsigma_wTOF[id]), t.p(), nsigma);
+          if (flagFourDemensionQA) {
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:434]:
```cxx
if (fillTHnSparses) {

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 3]
```diff
@@ -76,14 +79,17 @@ struct tpcPidQa {
...
+                                                            "wTOF/nsigma/eta/Tr", "wTOF/nsigma/eta/He", "wTOF/nsigma/eta/Al"};
 
   HistogramRegistry histos{"Histos", {}, OutputObjHandlingPolicy::AnalysisObject};
 
   Configurable<int> logAxis{"logAxis", 1, "Flag to use a log momentum axis"};
   Configurable<int> nBinsP{"nBinsP", 3000, "Number of bins for the momentum"};
   Configurable<float> minP{"minP", 0.01, "Minimum momentum in range"};
   Configurable<float> maxP{"maxP", 20, "Maximum momentum in range"};
-  ConfigurableAxis etaBins{"etaBins", {100, -1.f, 1.f}, "Binning in eta"};
+  ConfigurableAxis etaBins{"etaBins", {50, -1.f, 1.f}, "Binning in eta"};
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:92]: I think leave the default eta binning as it was, just we make sure to change the binning in the config when we run including the THnSparse

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 4]
```diff
@@ -175,6 +183,14 @@ struct tpcPidQa {
     const AxisSpec expSigmaAxis{expSigmaBins, Form("Exp_{#sigma}^{TPC}(%s)", pT[id])};
     histos.add(hexpsigma[id].data(), "", kTH2F, {pAxis, expSigmaAxis});
 
+    const AxisSpec etaAxis{etaBins, "#it{#eta}"};
+    const AxisSpec tpcnclsAxis{tpcNclsBins, "TPC #cls"};
+
+    HistogramConfigSpec defaultParticleHists{HistType::kTHnSparseF, {pAxis, etaAxis, nSigmaAxis, tpcnclsAxis}};
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:189]: defaultParticleHists -> particleSparseHists

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 5]
```diff
@@ -185,6 +201,12 @@ struct tpcPidQa {
     histos.add(hdelta_pt_neg_wTOF[id].data(), "With TOF Negative", kTH2F, {ptAxis, deltaAxis});
     histos.add(hexpsigma_wTOF[id].data(), "With TOF", kTH2F, {pAxis, expSigmaAxis});
     histos.add(hnsigma_wTOF[id].data(), Form("With TOF %s", axisTitle), kTH2F, {pAxis, nSigmaAxis});
+
+    HistogramConfigSpec defaultParticleHists_wTOF{HistType::kTHnSparseF, {pAxis, etaAxis, nSigmaAxis, tpcnclsAxis}};
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:205]: defaultParticleHists -> particleSparseHists

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 6]
```diff
@@ -175,6 +183,14 @@ struct tpcPidQa {
     const AxisSpec expSigmaAxis{expSigmaBins, Form("Exp_{#sigma}^{TPC}(%s)", pT[id])};
     histos.add(hexpsigma[id].data(), "", kTH2F, {pAxis, expSigmaAxis});
 
+    const AxisSpec etaAxis{etaBins, "#it{#eta}"};
+    const AxisSpec tpcnclsAxis{tpcNclsBins, "TPC #cls"};
+
+    HistogramConfigSpec defaultParticleHists{HistType::kTHnSparseF, {pAxis, etaAxis, nSigmaAxis, tpcnclsAxis}};
+    if (flagFourDemensionQA) {
+      histos.add(hnsigma_eta[id].data(), axisTitle, defaultParticleHists);
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:191]: defaultParticleHists -> particleSparseHists , hnsigma_eta -> hnsigma_p_eta_Ncl

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 7]
```diff
@@ -185,6 +201,12 @@ struct tpcPidQa {
     histos.add(hdelta_pt_neg_wTOF[id].data(), "With TOF Negative", kTH2F, {ptAxis, deltaAxis});
     histos.add(hexpsigma_wTOF[id].data(), "With TOF", kTH2F, {pAxis, expSigmaAxis});
     histos.add(hnsigma_wTOF[id].data(), Form("With TOF %s", axisTitle), kTH2F, {pAxis, nSigmaAxis});
+
+    HistogramConfigSpec defaultParticleHists_wTOF{HistType::kTHnSparseF, {pAxis, etaAxis, nSigmaAxis, tpcnclsAxis}};
+    if (flagFourDemensionQA) {
+      histos.add(hnsigma_eta_wTOF[id].data(), Form("With TOF %s", axisTitle), defaultParticleHists_wTOF);
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:207]: defaultParticleHists -> particleSparseHists , hnsigma_eta -> hnsigma_p_eta_Ncl

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 8]
```diff
@@ -76,14 +79,17 @@ struct tpcPidQa {
   static constexpr std::string_view hsignal_wTOF[Np] = {"wTOF/signal/El", "wTOF/signal/Mu", "wTOF/signal/Pi",
                                                         "wTOF/signal/Ka", "wTOF/signal/Pr", "wTOF/signal/De",
                                                         "wTOF/signal/Tr", "wTOF/signal/He", "wTOF/signal/Al"};
+  static constexpr std::string_view hnsigma_eta_wTOF[Np] = {"wTOF/nsigma/eta/El", "wTOF/nsigma/eta/Mu", "wTOF/nsigma/eta/Pi",
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:82]: hnsigma_eta -> hnsigma_p_eta_Ncl
nsigma/eta (in path) --> nsigma/sparsePinEtaNcl

File: DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx [hunk 9]
```diff
@@ -50,6 +50,9 @@ struct tpcPidQa {
   static constexpr std::string_view hnsigma_pt[Np] = {"nsigma/pt/El", "nsigma/pt/Mu", "nsigma/pt/Pi",
                                                       "nsigma/pt/Ka", "nsigma/pt/Pr", "nsigma/pt/De",
                                                       "nsigma/pt/Tr", "nsigma/pt/He", "nsigma/pt/Al"};
+  static constexpr std::string_view hnsigma_eta[Np] = {"nsigma/eta/El", "nsigma/eta/Mu", "nsigma/eta/Pi",
```
Comment (jezwilkinson) [inline DPG/Tasks/AOTTrack/PID/qaPIDTPC.cxx:53]: hnsigma_eta -> hnsigma_p_eta_Ncl
nsigma/eta (in path) --> nsigma/sparsePinEtaNcl

**Review by jezwilkinson — CHANGES_REQUESTED** [review]
Hi @chengtt0406 , from the tests we did at GSI the content of the histograms is fine and I think the memory usage should be OK (we still need to check how this scales on Grid, but the default is that they are not stored so it shouldn't be disruptive when they aren't requested)
Looking at the code in detail I think it'd be better to rename some of the variables to be a bit more descriptive (e.g. "nsigma/eta" in the path should be something more like "nsigma/sparsePinEtaNcl" so that it's clearer what it actually contains)

**Review by njacazio — APPROVED** [review]
Perfect thanks!

**Pull request merged**